modules/weaver-tests/src/smithy4s/weavertests/WeaverTests.scala

```diff
@@ -0,0 +1,108 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.weavertests
+
+import cats.effect.IO
+import org.http4s.HttpApp
+import org.http4s.HttpRoutes
+import org.http4s.Uri
+import smithy.test._
+import smithy4s.Document
+import smithy4s.Endpoint
+import smithy4s.Service
+import smithy4s.ShapeTag
+import smithy4s.UnsupportedProtocolError
+import weaver._
+
+case class GeneratedTest(name: String, assertions: IO[Expectations])
+
+abstract class WeaverTests[P, Alg[_[_, _, _, _, _]], Op[_, _, _, _, _]](
+    protocolTag: ShapeTag[_],
+    client: (
+        HttpApp[IO],
+        Uri
+    ) => Either[UnsupportedProtocolError, smithy4s.Monadic[Alg, IO]],
+    server: smithy4s.Monadic[Alg, IO] => Either[
+      UnsupportedProtocolError,
+      HttpRoutes[IO]
+    ]
+)(implicit service: Service[Alg, Op])
+    extends SimpleIOSuite {
+
+  private class EndpointTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO]
+  )(implicit ce: CompatEffect) {
+
+    def generateTests() = {
+      requestTests() ++ responseTests()
+    }
+
+    private def requestTests(): List[GeneratedTest] = {
+      val inputFromDocument = Document.Decoder.fromSchema(endpoint.input)
+      val whrtc = new WeaverHttpRequestTestCase(client, server)
+
+      val testCases =
+        endpoint.hints
+          .get(HttpRequestTests)
+          .map(_.value)
+          .getOrElse(Nil)
+          .filter(_.protocol == protocolTag.id.toString())
+
+      val clientTests = appliesTo(testCases)(AppliesTo.CLIENT, _.appliesTo)
+        .map(whrtc.makeClientTest(endpoint, _, inputFromDocument))
+
+      val serverTests = appliesTo(testCases)(AppliesTo.SERVER, _.appliesTo)
+        .map(whrtc.makeServerTest(endpoint, _, inputFromDocument))
+
+      serverTests ++ clientTests
+    }
+
+    private def responseTests(): List[GeneratedTest] = {
+      val inputFromDocument = Document.Decoder.fromSchema(endpoint.output)
+      val whrtc = new WeaverHttpResponseTestCase(client, server)
+
+      val testCases =
+        endpoint.hints
+          .get(HttpResponseTests)
+          .map(_.value)
+          .getOrElse(Nil)
+          .filter(_.protocol == protocolTag.id.toString())
+
+      val clientTests = appliesTo(testCases)(AppliesTo.CLIENT, _.appliesTo)
+        .map(whrtc.makeClientTest(endpoint, _, inputFromDocument))
+
+      val serverTests = appliesTo(testCases)(AppliesTo.SERVER, _.appliesTo)
+        .map(whrtc.makeServerTest(endpoint, _, inputFromDocument))
+
+      serverTests ++ clientTests
+    }
+
+    private def appliesTo[TC](testCases: List[TC])(
+        value: AppliesTo,
+        appliesTo: TC => Option[AppliesTo]
+    ): List[TC] =
+      testCases.collect {
+        case tc if appliesTo(tc).forall(_ == value) => tc
+      }
+  }
+
+  def recordTests(implicit ce: CompatEffect) = {
```

Comment:
I think this should be applied at construction without the user having to call it.